All in one operator rebase merge to master

Jeffwan opened the pull request:
Phase I (#1298 ) and Phase II (#1299 ) work are done and `all-in-one-operator` branch pass e2e test. 
Original tensorflow controller can coexist with refactored operator.

Since we won't have massive refactor work. I think it's a good time to merge back to master. 
Phase III work can be done against master laster https://github.com/kubeflow/tf-operator/issues/1318

What do you think? 

/cc @kubeflow/wg-training-leads

zw0610 committed:
Add XGBoost API (#1286)

Jeffwan committed:
Temporarily add Jeffwan@ to OWNERS (#1287)

Jeffwan committed:
Generate tfjob 1.19.x compatible clientset (#1290)

* Resolve GroupVersion redeclared issue

`go build` will fail since same group version is redeclared in groupversion_info.go

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Generate tfjob 1.19.x compatible clientset

1. Update dependencies to 1.19.x compatible
2. Update update-codegen.sh to use pkg from go modules pkg
3. Generate clientsets which supports context

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan committed:
Fix all client request that needs contexts (#1292)

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

zw0610 committed:
add pytorch API and controller (#1294)

Jeffwan committed:
Add XGBoost controller (#1293)

* Add XGBoost controller

1. Move major controller logics from kubeflow/xgboost-operator to this repo.
2. Adapt to kubebuilder 3.0.0 change
3. Add xgboost train example
4. Leave some TODOs for code refactor later

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Move examples from config to example/ folder

zw0610 committed:
add tf reconciler to the unified operator (#1295)

* add tensorflow api

* check in tensorflow reconciler

fix issues from comment

Jeffwan committed:
Add MXJob v1 api and controller (#1296)

- Generate MXJob api and manifests
- Rewrite MXNet controller logic in reconciler pattern

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan committed:
Extract reusable codes to common utility (#1297)

* Remove duplicate FakeWorkQueue

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Extract utility to common

Move ConvertPodList, ConvertServiceList to common library

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Move gang scheduing setting to common

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Extract OnDependentCreate/Delete handler to utils

Refactor follow methods to job independent and move to common utils
- OnDependentCreateFunc
- OnDependentDeleteFunc

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Refactor SatisfiedExpectations to a common method

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan committed:
Unify code structure of training job api (#1300)

* Add docs, register in job apis

- Clean up register.go and use reference in groupversion_info.go (generated by kubebuilder) instead
- Add doc.go and register.go for framework missing them
- Remove global tag “+k8s:deepcopy-gen=package,register” and this should be taken care of by controller-gen.

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Generate defaults and openapi for all frameworks

1. Update codegen scripts to generate defaulters and openapi spec for all frameworks
2. Create defaults.go for framework which miss it.

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Add Kind in constants.go

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Update update-codegen to generate openapi-gen

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Run goimports to format the files

* Add Plural and Singular to apis constants.go

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan committed:
Reorganize controllers code structure (#1302)

* Fix crd installation issue

CustomResourceDefinition.apiextensions.k8s.io "tfjobs.kubeflow.org" is invalid:
spec.validation.openAPIV3Schema.properties[metadata]: Forbidden: must not specify anything other than name and generateName,
but metadata is implicitly specified

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Clean up controllers

1. Move all ControllerInterface implementations to [framework]_controller.go
2. Clean up job.go pod.go service.go status.go
3. Follow exact same method order here https://github.com/kubeflow/common/blob/9ffa565bc60e08936f7f80cb3f491cf53f256e7f/pkg/apis/common/v1/interface.go

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Embed common.JobController instead of TFController

I notice previous change embed TFController which means most of the methods still use tfjob client. We need to make controller consistent and `TFJobReconciler` should override `common.JobController`.

With this change, it would be very easy to make original tf operator and new reconciler work together. We also avoid one more refactor when we will remove original TFController codes

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan committed:
Use low-level controller and handlers in SetupWithManager (#1315)

* Use low-level controller and handlers in SetupWithManager

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Add job validation in reconcile loop

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Set defaults in onOwnerCreateFunc

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Correctly update job status in apiserver

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Remove Flag for Kubeconfig to fix flag redefined

CI reports `flag redefined: kubeconfig` issue and this is due to duplicate flag registration. See https://github.com/kubeflow/tf-operator/issues/1316 for more details.

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

* Fix tensorflow job missing default port issue

This is to fix original controller issue. It leverages init to register default to scheme. In our unversal operator project, we did clean up for register.go and break the case.

For more details, please check https://github.com/kubeflow/tf-operator/issues/1317#issuecomment-890397705

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

Jeffwan commented:
/hold

Jeffwan commented:
>

> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.

@zw0610 Can you help on this?

terrytangyuan reviewed (approved):
LGTM and +1 to merge manually

zw0610 commented:
@googlebot I consent.

zw0610 commented on config/rbac/tfjob_editor_role.yaml:
```diff
@@ -0,0 +1,24 @@
+  resources:
+  - tfjobs/status
+  verbs:
+  - get
```
do we need to add `update`, `patch` to resources `tfjobs/status` as well as mxnetjobs, pytorchjobs and xboostjobs?

Jeffwan replied:
I remember we use same tool to generate editor roles.  Do they look different now?

Jeffwan replied:
@zw0610 seems rest of the role are same.  update, patch are not part of the editor role. 

BTW, this is different from the role running the operator. I create a role to use for operator in #1325

zw0610 replied:
SGTM

gaocegege commented:
/cc @kubeflow/wg-automl-leads

Jeffwan commented:
Note, we don't bring **new** changes to master. All the commits in this PR has been approved and merged in `all-in-one-operator` branch.  If you think this change is too big, feel free to click original PR link to check background. This may help you expedite the review process.

![image](https://user-images.githubusercontent.com/4739316/127970204-d9dcf9f5-90fb-4f39-ac03-763227baae88.png)

gaocegege commented:
SGTM

Jeffwan commented:
/cc @johnugeorge  @andreyvelich  any comments?

andreyvelich commented:
Sounds good for the changes @Jeffwan.

> Original tensorflow controller can coexist with refactored operator.

Since we are making changes for the tf controller, do we still need to build separate image for tf-operator controller or we will use common `kubeflow-training` image once we merge this PR ?

andreyvelich commented on Makefile:
```diff
@@ -0,0 +1,108 @@
+run: manifests generate fmt vet ## Run a controller from your host.
+	go run ./main.go
+
+docker-build: test ## Build docker image with the manager.
```
Where we will locate Dockerfile for `training-operator` image ?

Jeffwan replied:
https://github.com/kubeflow/tf-operator/pull/1325 ./build/images/training-operator/Dockerfile

Maybe we should merge that PR first

andreyvelich replied:
@Jeffwan Does it look good to create separate dir only for one `Dockerfile` or we are planing to have some other `Dockerfiles` under `./build/image` ?
For examples we store `Dockerfile` in the same dir where example is located: https://github.com/kubeflow/tf-operator/tree/8430a040ad38f3ea1f2fd4b3f077d0871b4cdd52/examples/tf_sample.

Jeffwan replied:
You mean something like `Dockerfile.tf` and `Dockerfile.training` in the same folder?  I think that make sense. I can document this change and make some changes later.

andreyvelich replied:
Yes, we can do that.
Or we can move `Dockerfile` to `cmd/training-operator.v1/Dockerfile` similar to `main.go`.

Jeffwan replied:
Got your point. Currently, I follow existing way to organize the files. I think these be changes for sure.  Originally, we put Dockerfile and main.go in separate folders. We can consider to put them together later. I think create an improvement issue to track it.  https://github.com/kubeflow/tf-operator/issues/1333

```
build
└── images
    ├── tf_operator
    │   └── Dockerfile
cmd
├── tf-operator.v1
│   ├── app
│   │   ├── options
│   │   │   └── options.go
│   │   └── server.go
│   └── main.go

```

andreyvelich replied:
Thanks @Jeffwan!

andreyvelich commented on Makefile:
```diff
@@ -0,0 +1,108 @@
+##@ Deployment
+
+install: manifests kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build config/crd | kubectl apply -f -
```
Will we use `/config` or `/manifests` for our final manifests ?

Jeffwan replied:
config is generated from kubebuilder. I have not moved all manifest to `/manifests` yet.

andreyvelich replied:
Got it, thank you for clarification!

andreyvelich commented on Makefile:
```diff
@@ -0,0 +1,108 @@
+manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./pkg/apis/..." output:crd:artifacts:config=config/crd/bases
+
+generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
```
Do we need these instructions since [`update-codegen.sh`](https://github.com/kubeflow/tf-operator/blob/43d950df3b75271ae2b77ef5a7c291c4c21cf7d8/hack/update-codegen.sh) covers it, or I am missing something ?

Jeffwan replied:
`controller-gen` is built out of different generators. I would love to keep this Make file and it helps to generate crds and rbacs.

andreyvelich commented on main.go:
```diff
@@ -0,0 +1,123 @@
+/*
```
I think we can move this file to `cmd/main.go` within Dockerfile in the future.
What do you think @Jeffwan ?

Jeffwan replied:
Sounds good. `cmd/training-operator.v1/main.go`

Jeffwan replied:
Hi Andrey, I include this change in a separate PR. This PR includes most changes we made in last two stages. There's follow up PR coming to address these suggestions.

andreyvelich commented on pkg/apis/mxnet/v1/groupversion_info.go:
```diff
@@ -0,0 +1,34 @@
+// Copyright YEAR The Kubeflow Authors
```
Will we modify script to change `YEAR` to `2021` in the future ?

Jeffwan replied:
Yeah, I document this issue here https://github.com/kubeflow/tf-operator/issues/1309.

andreyvelich replied:
Thanks!

andreyvelich commented on pkg/client/clientset/versioned/clientset.go:
```diff
@@ -57,7 +57,7 @@ func NewForConfig(c *rest.Config) (*Clientset, error) {
 	configShallowCopy := *c
```
We need to remove `clientset` in the future once we finish implementation.

Jeffwan replied:
This need more discussion. Do we want this repo to host all client sets or a separate repo? 

We are not sure which project interact with training crds. If there's any, we should prioritize the story

andreyvelich replied:
Sounds good, let's track it somewhere.

Jeffwan replied:
I create an issue here https://github.com/kubeflow/tf-operator/issues/1332 to track this discussion

Jeffwan committed:
Remove inline runPolicy and register defaulter (#1330)

Jeffwan commented:
I think I addressed all the concerns and suggestions. Some of the improvements are not made in this PR.  This PR won't include any direct changes. This is a rebase change from `all-in-one-operator` to `master`. Any changes should be made against `all-one-one-operator` branch. 

For p1 issues, changes are included in https://github.com/kubeflow/tf-operator/pull/1325. 
For p2 issues, I create issues and they will resolve it later. 

This is a clean PR to make original test work. It's safe to merge it separately.
